### PR TITLE
LTI and SAML: The Basics

### PR DESCRIPTION
**Pull request by gera-cameron**
Story details: https://app.clubhouse.io/itprotv/story/18732

Diagrams to come. Wanted to get some eyes on the text :)

### CODE REVIEW

**Review by codygman — APPROVED** [review]

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 1]
```diff
@@ -0,0 +1,96 @@
+# LTI and SAML: The Basics
+
+by Cameron Gera (@gera-cameron)
+
+LTI and SAML single sign-on strategies are incredibly useful, but they are some of the hardest things to grok so I am going to put it in terms that normal developers and system administrators understand.
+These two standards have full-fledged specs out there on the internet and if you work well with those standardized papers I would you encourage you to check those out.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
These two standards have full-fledged specs out there on the internet and if you work well with those standardized papers I would encourage you to check those out.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+Sadly, every time I read those I left more confused and lost, so I am going to take a different route.
+What was most useful for me were diagrams so as I get into the details about each I will provide diagrams to explain what I am saying.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
What was most useful for me were diagrams. As I get into the details about each I will provide diagrams to explain what I am saying. So if you are looking for a simplified version of what these two standards are, then stick around.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+So if you are looking for a simplified version of what these two standards are, then stick around.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion

```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+## LTI Standard
+
+The LTI standard is very common in the academic space and somewhat in the business space.
+LTI standard stands for Learning Tools Interoperability.
+Knowing this name isn't super useful besides letting you know that a learning management system (LMS) tends to integrate via this strategy.
+Tools like Canvas, Moodle, and Blackboard are examples of LMS.
+
+### How it works
+
+So for those of you who are junkies for knowing how every little thing about how stuff works, this is for you.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]: I'm having a bit of a hard time reading this sentence fluently, unsure what to change.

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+Although I am not an expert on LTI, I will give my best effort for you.
+
+![LTI Diagram][]
+
+Okay, so we have two entities that will be communicating with each other.
+One is the LTI Consumer and the LTI Tool Provider.
+The LTI Consumer is the LMS.
+The LTI Tool Provider is the external service that the LMS is trying to access.
+The authentication works through a key and secret system with some other configurable pieces on the LMS side.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]: What other configurable pieces? Is there a resource that can help me?

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+The terms to know in regards to authentication are the following:
+
+- **Consumer Key** - This is a shared key between the Consumer and the Tool Provider.
+- **Shared Secret** - This is the privately shared secret between the Consumer and the Tool Provider.
+- **Launch URL** - This is the URL that the user would go to when launching the Tool Provider.
+- **Domain** - This is derived from the Launch URL and seems to be optional in my experience.
+- **Privacy** - This is the level of privacy for the information shared with the Tool Provider.
+- **Custom Fields** - These are the custom fields that will allow the Tool Provider to tailor the experience to a specific Consumer.
+
+These variables are pretty simple to understand and they should be scary.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]: Why should they be scary?

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+A tough aspect is implementing the logic for authentication on the Tool Provider side.
+But fear not!
+There are libraries out there that will help like [PassportJS](http://www.passportjs.org/packages/passport-lti) and others to help.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
There are libraries out there that will help like [PassportJS](http://www.passportjs.org/packages/passport-lti)
```

Is it worth mentioning other libraries or a resource that may list them?

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+These libraries take care of some of the magic associated with LTI, but the give you a friendly interface to work with.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
These libraries take care of some of the magic associated with LTI, but they give you a friendly interface to work with.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+As you can see the user begins by accessing the service provider and informing them that they belong to a certain identity provider.
+Once the service provider knows who the user is and where to send them to be authorized, they are redirected to the IDP login screen.
+After the IDP says they are good, they redirect back to the TP with a set of user identifiable information.
+When the SP sees this, it verifies that the payload was signed with the same certificate that the SP has on file.
+Now that we have proven the user is from a trusted source, the SP can give access to the user.
+
+Some terms to know are the following:
+
+- **Entry Point** - This is the endpoint on the Identity Provider side that the Service Provider redirects to for authentication.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
- **Entry Point** - This is the endpoint on the IDP side that the SP redirects to for authentication.
```
I'm indiffirent on making them shorter or not but I would expect it to be consistent.

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+- **Callback URL** - This is an endpoint on the Service Provider side that listens for the SAML Payload from the Identity Provider.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
- **Callback URL** - This is an endpoint on the SP side that listens for the SAML Payload from the IDP.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+- **Issuer** - This is the Service Provider's registered domain.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
- **Issuer** - This is the SP's registered domain.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+- **Certificate** - This is what signs the payloads to verify security.
+- **Relay State** - This is a field that the Service Provider can set when redirecting to the entry point that will be sent back in the payload.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
- **Relay State** - This is a field that the SP can set when redirecting to the entry point that will be sent back in the payload.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+
+Sounds pretty straight forward right and it is.
```
Comment (aschmois) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
Sounds pretty straight forward right? and it is.
```

**Review by aschmois — APPROVED** [review]
Overall looks good, just a few typos.

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+## LTI Standard
+
+The LTI standard is very common in the academic space and somewhat in the business space.
+LTI standard stands for Learning Tools Interoperability.
+Knowing this name isn't super useful besides letting you know that a learning management system (LMS) tends to integrate via this strategy.
+Tools like Canvas, Moodle, and Blackboard are examples of LMS.
+
+### How it works
+
+So for those of you who are junkies for knowing how every little thing about how stuff works, this is for you.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
LTI is an SSO strategy that is always initiated from the users LMS. 
It is a fairly streamlined process and I'll do my best to keep it clear. 
Let's check out this diagram to visualize the flow.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+Although I am not an expert on LTI, I will give my best effort for you.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+The terms to know in regards to authentication are the following:
+
+- **Consumer Key** - This is a shared key between the Consumer and the Tool Provider.
+- **Shared Secret** - This is the privately shared secret between the Consumer and the Tool Provider.
+- **Launch URL** - This is the URL that the user would go to when launching the Tool Provider.
+- **Domain** - This is derived from the Launch URL and seems to be optional in my experience.
+- **Privacy** - This is the level of privacy for the information shared with the Tool Provider.
+- **Custom Fields** - These are the custom fields that will allow the Tool Provider to tailor the experience to a specific Consumer.
+
+These variables are pretty simple to understand and they should be scary.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
These variables are pretty simple to understand and they are all you need to set up when working with LTI.
If you get these, you'll be able to talk the lingo and have a good jist of whats going on.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+A tough aspect is implementing the logic for authentication on the Tool Provider side.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
The tough aspect is implementing the logic for authentication on the Tool Provider side.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+## LTI Standard
+
+The LTI standard is very common in the academic space and somewhat in the business space.
+LTI standard stands for Learning Tools Interoperability.
+Knowing this name isn't super useful besides letting you know that a learning management system (LMS) tends to integrate via this strategy.
+Tools like Canvas, Moodle, and Blackboard are examples of LMS.
+
+### How it works
+
+LTI is an SSO strategy that is always initiated from the users LMS.
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
LTI is an SSO strategy that is always initiated from the user's LMS. 
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+### How it works
+
+LTI is an SSO strategy that is always initiated from the users LMS. 
+It is a fairly streamlined process and I'll do my best to keep it clear. 
+Let's check out this diagram to visualize the flow.
+
+![LTI Diagram][]
+
+Okay, so we have two entities that will be communicating with each other.
+One is the LTI Consumer and the LTI Tool Provider.
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
One is the LTI Consumer and the other is the LTI Tool Provider.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+The LTI Consumer is the LMS.
+The LTI Tool Provider is the external service that the LMS is trying to access.
+The authentication works through a key and secret system with some other configurable pieces on the LMS side.
+
+The terms to know in regards to authentication are the following:
+
+- **Consumer Key** - This is a shared key between the Consumer and the Tool Provider.
+- **Shared Secret** - This is the privately shared secret between the Consumer and the Tool Provider.
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]: Is there a clear difference between the consumer key and the shared secret? Based on the wording I think the key might be public and the secret is private, but I'm not certain.

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+- **Launch URL** - This is the URL that the user would go to when launching the Tool Provider.
+- **Domain** - This is derived from the Launch URL and seems to be optional in my experience.
+- **Privacy** - This is the level of privacy for the information shared with the Tool Provider.
+- **Custom Fields** - These are the custom fields that will allow the Tool Provider to tailor the experience to a specific Consumer.
+
+These variables are pretty simple to understand and they are all you need to set up when working with LTI.
+If you get these, you'll be able to talk the lingo and have a good jist of whats going on.
+The tough aspect is implementing the logic for authentication on the Tool Provider side.
+But fear not!
+There are libraries out there that will help like [PassportJS](http://www.passportjs.org/packages/passport-lti)
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
There are libraries out there that will help like [PassportJS](http://www.passportjs.org/packages/passport-lti).
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 22]
```diff
@@ -0,0 +1,97 @@
...
+These libraries take care of some of the magic associated with LTI, but they give you a friendly interface to work with.
+We have been using PassportJS at work for quite a few years and we rarely have issues with our original implementation.
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
We have been using PassportJS at ITProTV for quite a few years and we rarely have issues with our original implementation.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 23]
```diff
@@ -0,0 +1,97 @@
...
+When setting up SAML, a set of metadata is shared between the two providers to set up a secure connection and make sure that all needed values for the SP are sent.
+See the flow diagram below for the end user's flow for both SP initiated and IDP initiated.
+
+![SAML SP Diagram][]
+
+![SAML IDP Diagram][]
+
+As you can see the user begins by accessing the service provider and informing them that they belong to a certain identity provider.
+Once the service provider knows who the user is and where to send them to be authorized, they are redirected to the IDP login screen.
+After the IDP says they are good, they redirect back to the TP with a set of user identifiable information.
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
After the IDP says they are good, they redirect back to the SP with a set of user identifiable information.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 24]
```diff
@@ -0,0 +1,97 @@
...
+Some terms to know are the following:
+
+- **Entry Point** - This is the endpoint on the IDP side that the SP redirects to for authentication.
+- **Callback URL** - This is an endpoint on the SP side that listens for the SAML Payload from the IDP.
+- **Issuer** - This is the SP's registered domain.
+- **Certificate** - This is what signs the payloads to verify security.
+- **Relay State** - This is a field that the SP can set when redirecting to the entry point that will be sent back in the payload.
+
+
+Sounds pretty straight forward right? and it is.
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
Sounds pretty straightforward right? and it is.
```

Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
Sounds pretty straight forward right? And it is!
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 25]
```diff
@@ -0,0 +1,97 @@
...
+SAML is one of the most common integrations that I am familiar with.
+The ability to use a single login to access many tools is becoming a must-have in the corporate world.
+I would encourage you to check out more documentation if you have more curiosity!
```
Comment (tfausak) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
I would encourage you to check out more documentation if you are curious!
```

**Review by tfausak — APPROVED** [review]

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 26]
```diff
@@ -0,0 +1,97 @@
...
+
+Okay, so we have two entities that will be communicating with each other.
+One is the LTI Consumer and the other is the LTI Tool Provider.
+The LTI Consumer is the LMS.
+The LTI Tool Provider is the external service that the LMS is trying to access.
+The authentication works through a key and secret system with some other configurable pieces on the LMS side.
+
+The terms to know in regards to authentication are the following:
+
+- **Consumer Key** - This is a shared key between the Consumer and the Tool Provider.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
- **Consumer Key** - This is a privately shared key between the Consumer and the Tool Provider and must match on both ends.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 27]
```diff
@@ -0,0 +1,97 @@
...
+- **Shared Secret** - This is the privately shared secret between the Consumer and the Tool Provider.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
- **Shared Secret** - This is the privately shared secret between the Consumer and the Tool Provider and must match on both ends.
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown [hunk 28]
```diff
@@ -0,0 +1,97 @@
...
+Some terms to know are the following:
+
+- **Entry Point** - This is the endpoint on the IDP side that the SP redirects to for authentication.
+- **Callback URL** - This is an endpoint on the SP side that listens for the SAML Payload from the IDP.
+- **Issuer** - This is the SP's registered domain.
+- **Certificate** - This is what signs the payloads to verify security.
+- **Relay State** - This is a field that the SP can set when redirecting to the entry point that will be sent back in the payload.
+
+
+Sounds pretty straightforward right? and it is.
```
Comment (gera-cameron) [inline _posts/2019-08-23-lti-and-saml-the-basics.markdown]:
```suggestion
Sounds pretty straightforward right? And it is!
```

File: _posts/2019-08-23-lti-and-saml-the-basics.markdown (hunk 5 above)
Comment (gera-cameron) [reply _posts/2019-08-23-lti-and-saml-the-basics.markdown]: The resource is below.

**Review by tfausak — APPROVED** [review]

**Pull request merged**